<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting Started with <code>polyIBD</code> • polyIBD</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started with &lt;code&gt;polyIBD&lt;/code&gt;">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">polyIBD</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.8.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/polyIBD.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/nickbrazeau/polyIBD">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Getting Started with <code>polyIBD</code>
</h1>
                        <h4 class="author">Nicholas F. Brazeau &amp; Robert Verity</h4>
            
            <h4 class="date">May 31, 2018</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/nickbrazeau/polyIBD/blob/master/vignettes/polyIBD.Rmd"><code>vignettes/polyIBD.Rmd</code></a></small>
      <div class="hidden name"><code>polyIBD.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p><code>polyIBD</code> is an R package that can be used to infer regions of identity by descent (IBD) from samples that are unphased and potentially polyclonal.</p>
<p>Recently, several programs have been written to address this problem, including Henden/Bahlo’s <a href="https://github.com/bahlolab/isoRelate">isorelate</a> and Schaffner/Neafsey’s <a href="https://github.com/glipsnort/hmmIBD">hmmIBD</a>. Here we extend upon these approaches and <strong>assume that the multiplicity of infection (MOI) is unkown and is not limited to 2.</strong></p>
<p>Using a bayesian approach, the main strength of <code>polyIBD</code> is that MOI estimates are unbounded and can be inferred directly along with other model parameters. As such, <code>polyIBD</code> has the ability to infer:</p>
<ol style="list-style-type: decimal">
<li>Regions of IBD</li>
<li>MOI estimates for both samples<br>
</li>
<li>Genetic Relatedness</li>
<li>Rate of Recombination</li>
</ol>
<div id="description" class="section level2">
<h2 class="hasAnchor">
<a href="#description" class="anchor"></a>Description</h2>
<p><code>polyIBD</code> is a first-order Hidden Markov Model (HMM) that uses the Metropolis-Hasting Algorithm to infer model parameters. Within the HMM framework, <code>polyIBD</code> assumes that the transition probabilities and emission probabilites can be described and uses the forward-backward algorithm to calculate the likelihood of the hidden state. For an extended description of the model, please view the <a href="">supplementary information</a>.</p>
</div>
</div>
<div id="getting-started-with-polyibd" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-started-with-polyibd" class="anchor"></a>Getting Started with <code>polyIBD</code>
</h1>
<p>First you will need to have successfully installed <code>polyIBD</code> as described previously in the <a href="https://github.com/nickbrazeau/polyIBD">README</a>.</p>
<div id="input" class="section level2">
<h2 class="hasAnchor">
<a href="#input" class="anchor"></a>Input</h2>
<p>The input for <code>polyIBD</code> is a filtered<a href="#fn1" class="footnoteRef" id="fnref1"><sup>1</sup></a> Variant Call File (VCF) that follows the <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF specifciations</a> for the Genotype Filed and has a genotype (GT) value for each sample. In additon, VCFs must be unphased and called as a diploid. As such, acceptable GT values are: 0/0, 0/1, 1/1, and “.”, for homozygous referent, heterozygous, homozygous alternative, and missing, respectively. Currently <code>polyIBD</code> does not handle missing data, which is a limitation that will be resolved shortly.</p>
</div>
<div id="simulated-data" class="section level2">
<h2 class="hasAnchor">
<a href="#simulated-data" class="anchor"></a>Simulated Data</h2>
<p>For the purpose of this vignette, we will simulate a VCF with two samples and use it to run <code>polyIBD</code>. First let’s input some specifications for the number of loci, the genetic relatedness, and the recombination rate for these samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># setup</span>
n &lt;-<span class="st"> </span><span class="fl">1e3</span> <span class="co"># this is the number of loci we would like to simulate</span>
rho_true &lt;-<span class="st"> </span><span class="fl">5e-3</span> <span class="co"># this is the recombination rate we are trying to infer</span>
f_true &lt;-<span class="st"> </span><span class="fl">0.2</span> <span class="co"># this is the proportion of genetic relatedness we are trying to infer </span>
m_true &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">3</span>) <span class="co"># this is the multiplicity of infection for sample1 and sample2, respectfully, that we are trying to infer</span>
pos &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">sample</span>(<span class="dv">10</span><span class="op">*</span>n, n)) <span class="co"># simulate some positions/genomic coordinates</span></code></pre></div>
<p>We will now call the built-in simulation function from <code>polyIBD</code> to take our inputs a simulate a two-sampled VCF. The simulation is using a beta distribution with the shape parameters alpha=1 and beta=1 to simiulate population allele frequencies. This distribution can be adjusted by the user but the 1,1 shape parameters provide a population with a high proportion of fixated alleles (to see the distribution, input the following command to your R Console: #&gt; hist(rbeta(1000, 0.5, 0.5))).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">44</span>)
<span class="co"># run the simulation </span>
sim &lt;-<span class="st"> </span><span class="kw"><a href="../reference/simData.html">simData</a></span>(<span class="dt">pos=</span><span class="kw">list</span>(<span class="dt">contig1=</span>pos,<span class="dt">contig2=</span>pos), 
               <span class="dt">m1=</span>m_true[<span class="dv">1</span>], <span class="dt">m2=</span>m_true[<span class="dv">2</span>], 
               <span class="dt">f=</span>f_true, <span class="dt">rho=</span>rho_true, <span class="dt">p=</span><span class="ot">NULL</span>, <span class="dt">p_shape1=</span><span class="fl">0.1</span>, <span class="dt">p_shape2=</span><span class="fl">0.1</span>)</code></pre></div>
<p>The ouptut of the <code>simData</code> function is a list of two lists and two dataframes:<br>
1. Simulated population major allele frequencies for both samples (<code>p</code>) 2. The haploid genotypes (0 for referrent allele, 2 for alternate allele) for both samples (<code>haploid</code>) 3. The haploid genotypes (0 for referrent allele, 2 for alternate allele) for both samples (<code>haploid</code>) 4. 4. The IBD genotypes at each positions genotypes (0 for referrent allele, 2 for alternate allele) for both samples (<code>haploid</code>) ⋅⋅* This is a main assumption in the package…TBD …</p>
<p align="center">
<img src="https://github.com/nickbrazeau/polyIBD/figures/IBD_match_sim.png" alt="ibd_match_plot"></p>
<!-- ## Running the MCMC  -->
<!-- ```{r, echo=T, eval=T} -->
<!-- # run MCMC -->
<!-- rho_max <- rho_true*10000 # this is user-specified -->
<!-- ret <- runMCMC(vcf=sim$vcf, p=sim$p, rho_max=rho_max, -->
<!--                burnin=5e3, samples=1e4, reportIteration=NA) # turning off report iteration for space -->
<!-- ``` -->
<!-- From the MCMC, a polyIBD class object is returned that contians both summary statistics and raw data. The summary statistics and raw data can be interrogated directly.  -->
<!-- ### Summary Data -->
<!-- Summary of the user-specified parameters for the MCMC burn-in and sampling iterations, acceptance rate (*ideally should be around 23%*), and the run-time of the function.    -->
<!-- In addition, posterior estimates with median and 95% confidence intervals provided.  -->
<!-- ```{r, echo=T, eval=T} -->
<!-- ret # is equivalent to summary(ret) -->
<!-- ``` -->
<!-- ### Raw Data -->
<!-- All of the raw data used to generate the above summary statistics is available within the returned MCMC object.  -->
<!-- ```{r, echo=T, eval=T} -->
<!-- summary(ret$raw) # is equivalent to summary(ret) -->
<!-- ``` -->
<!-- ## Viewing the Results -->
<!-- ### Base R -->
<!-- ```{r, echo=T, eval=T} -->
<!-- # plot using base R  -->
<!-- par(mfrow=c(2,2)) # for 4 plots -->
<!-- par(mar=c(1,3,3,1)) # you can adjust margins as you see fit -->
<!-- plot_m1(ret, ylim=c(0,5)) -->
<!-- abline(h=m_true[1], col=2) -->
<!-- plot_m2(ret, ylim=c(0,5)) -->
<!-- abline(h=m_true[2], col=2) -->
<!-- plot_f(ret) -->
<!-- abline(h=f_true, col=2) -->
<!-- plot_rho(ret) -->
<!-- abline(h=rho_true, col=2) -->
<!-- par(mfrow=c(1,1)) -->
<!-- par(mar=c(2.5,1,1,2.5)) -->
<!-- plot_IBD(ret) -->
<!-- ``` -->
<!-- ### Using `ggplot` Objects -->
<!-- ```{r, echo=T, eval=T} -->
<!-- # plot using ggplot from tidyverse -->
<!-- plotm1 <- polyIBD::ggplot_m1(ret) -->
<!-- plotm1 <- plotm1 + geom_hline(yintercept=m_true[1], colour="#cb181d") -->
<!-- plotm2 <- polyIBD::ggplot_m2(ret) -->
<!-- plotm2 <- plotm2 + geom_hline(yintercept=m_true[2], colour="#cb181d") -->
<!-- plotf <- polyIBD::ggplot_f(ret) -->
<!-- plotf <- plotf + geom_hline(yintercept=f_true, colour="#cb181d") -->
<!-- plotrho <- polyIBD::ggplot_rho(ret) -->
<!-- plotrho <- plotrho + geom_hline(yintercept=rho_true, colour="#cb181d") -->
<!-- ## convert plots to gtable objects -->
<!-- grobm1 <- ggplot2::ggplotGrob(plotm1) -->
<!-- grobm2 <- ggplot2::ggplotGrob(plotm2) -->
<!-- grobf <- ggplot2::ggplotGrob(plotf) -->
<!-- grobrho <- ggplot2::ggplotGrob(plotrho) -->
<!-- gridExtra::grid.arrange(grobm1, -->
<!--              grobm2, -->
<!--              grobf, -->
<!--              grobrho, -->
<!--              ncol=2) -->
<!-- ggplot_IBD(ret) -->
<!-- ``` -->
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Filtering will not be discussed here but resources include: <a href="https://gatkforums.broadinstitute.org/gatk/discussion/2806/howto-apply-hard-filters-to-a-call-set">GATK Forums</a>, the <a href="https://cran.r-project.org/web/packages/vcfR/vignettes/intro_to_vcfR.html">vcfR</a> tool, among others.<a href="#fnref1">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#overview">Overview</a><ul class="nav nav-pills nav-stacked">
<li><a href="#description">Description</a></li>
      </ul>
</li>
      <li>
<a href="#getting-started-with-polyibd">Getting Started with <code>polyIBD</code></a><ul class="nav nav-pills nav-stacked">
<li><a href="#input">Input</a></li>
      <li><a href="#simulated-data">Simulated Data</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Nicholas F Brazeau, Robert J Verity.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
